service-impl/cache-service-impl/simple-cache-service/src/main/java/uk/gov/gchq/palisade/cache/service/impl/CacheMetadata.java

```diff
@@ -0,0 +1,154 @@
+
+import java.util.Arrays;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Metadata information about a cache entry.
+ */
+public class CacheMetadata {
+
+    /**
+     * If true, then this cache entry has been retrieved from a local cache.
+     */
+    private final boolean wasRetrievedLocally;
+
+    /**
+     * If true, then this cache entry can be stored in a local cache.
+     */
+    private final boolean canRetrieveLocally;
+
+    /**
+     * Create a new metadata object.
+     *
+     * @param wasRetrievedLocally if the most recent retrieval came from the local cache or not
+     * @param canRetrieveLocally  whether the corresponding entry can be safely cached locally
+     */
+    public CacheMetadata(final boolean wasRetrievedLocally, final boolean canRetrieveLocally) {
+        this.wasRetrievedLocally = wasRetrievedLocally;
+        this.canRetrieveLocally = canRetrieveLocally;
+    }
+
+    /**
+     * Encodes extra metadata needed about this entry into the start of the byte array given to the backing store.
+     *
+     * @param encodedValue the encoded cache value object
+     * @param request      the cache request
+     * @return a new byte array containing metadata
+     */
+    public static byte[] addMetaData(final byte[] encodedValue, final AddCacheRequest<?> request) {
+        requireNonNull(encodedValue, "encodedValue");
+        requireNonNull(request, "request");
+        byte[] withMeta = new byte[encodedValue.length + 1];
+        System.arraycopy(encodedValue, 0, withMeta, 1, encodedValue.length);
+        if (request.getLocallyCacheable()) {
+            withMeta[0] = 1;
+        }
+        return withMeta;
+    }
+
+    /**
+     * Creates a new cache object with populated metadata based on the value byte array in the given cache object. The new
+     * cache object will have a populated metadata object and a new value byte array with the metadata removed.
+     *
+     * @param remoteRetrieve the cache object from the backing store
+     * @return a new cache object
+     * @throws IllegalStateException if the metadata is already populated in {@code remoteRetrieve}, or if no cache
+     *                               value is present
+     */
+    public static SimpleCacheObject populateMetaData(final SimpleCacheObject remoteRetrieve) {
+        requireNonNull(remoteRetrieve, "remoteRetrieve");
+        if (remoteRetrieve.metadata().isPresent()) {
+            throw new IllegalStateException("metadata already present");
+        }
+
+        byte[] withMeta = remoteRetrieve.getValue().orElseThrow(() -> new IllegalStateException("no cache value present"));
+        //remove the metadata from array
+        byte[] value = Arrays.copyOfRange(withMeta, 1, withMeta.length);
+
+        CacheMetadata metadata = new CacheMetadata(false, withMeta[0] == 1);
+        return new SimpleCacheObject(remoteRetrieve.getValueClass(), Optional.of(value), Optional.of(metadata));
```

Comment:
I think the SimpleCacheObject should be made mutable with the value and metadata being updated rather than create a new object, otherwise this part will become a performance bottleneck

Reply:
Done